File: apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/ResponseFetcher.java
package com.apollographql.apollo.fetcher;

import com.apollographql.apollo.interceptor.ApolloInterceptor;
import com.apollographql.apollo.internal.ApolloLogger;

/**
 * A ResponseFetcher is an {@link ApolloInterceptor} inserted at the beginning of a request chain.
 * It can control how a request is fetched by configuring {@link com.apollographql.apollo.interceptor.FetchOptions}.
 *
 * See {@link ApolloResponseFetchers} for a basic set of fetchers.
 */
public interface ResponseFetcher {

  /**
   * @param logger A {@link ApolloLogger} to log relevant fetch information.
   * @return The {@link ApolloInterceptor} that executes the fetch logic.
   */
  ApolloInterceptor provideInterceptor(ApolloLogger logger);

}

File: apollo-runtime/src/main/java/com/apollographql/apollo3/fetcher/ResponseFetcher.kt
package com.apollographql.apollo3.fetcher

import com.apollographql.apollo3.api.internal.ApolloLogger
import com.apollographql.apollo3.interceptor.ApolloInterceptor

/**
 * A ResponseFetcher is an [ApolloInterceptor] inserted at the beginning of a request chain.
 * It can control how a request is fetched by configuring [com.apollographql.apollo3.interceptor.FetchOptions].
 *
 * See [ApolloResponseFetchers] for a basic set of fetchers.
 */
interface ResponseFetcher {
  /**
   * @param logger A [ApolloLogger] to log relevant fetch information.
   * @return The [ApolloInterceptor] that executes the fetch logic.
   */
  fun provideInterceptor(logger: ApolloLogger?): ApolloInterceptor
}